coldfront/core/department/views.py

```diff
@@ -0,0 +1,47 @@
+from django.contrib.auth.mixins import LoginRequiredMixin
+from django.views.generic.edit import FormView
+from django.shortcuts import render
+from django.urls import reverse
+
+from coldfront.core.department.models import UserDepartment
+from coldfront.core.department.forms import DepartmentSelectionForm
+from coldfront.core.utils.common import import_from_settings
+
+# Create your views here.
+
+class UpdateDepartmentsView(LoginRequiredMixin, FormView):
+
+    form_class = DepartmentSelectionForm
+    template_name = 'user_update_departments.html'
+    login_url = '/'
+
+    error_message = 'Unexpected failure. Please contact an administrator.'
+
+    def form_valid(self, form):
+        user = self.request.user
+        form_data = form.cleaned_data
+        new_departments = form_data['departments']
+        userprofile = user.userprofile
+        for department in new_departments:
+            UserDepartment.objects.get_or_create(userprofile=userprofile,
```

Comment:
I added myself to "OTHER" using this form, then manually set it to `is_authoritative` via the shell, and re-submitted the form with "OTHER". An error was raised:

```
IntegrityError at /user/update-departments
duplicate key value violates unique constraint "department_userdepartmen_userprofile_id_departmen_ea98db45_uniq"
DETAIL:  Key (userprofile_id, department_id)=(1900, 1) already exists.
```

Generally, this error will be raised if the user tries to add a department they already have "authoritatively".

In the form, I'd probably not include authoritative associations in the initial form data. I'd list the ones they already have authoritatively on this form in a read-only field or table. If they try to add something they already have authoritatively, either raise an error or ignore it on the backend.

Reply:
I used default in `get_or_create`, which ignores departments that already exist authoritatively:
`for department in data['departments']:
                    UserDepartment.objects.get_or_create(userprofile=pi_profile,
                        department=department,
                        defaults={'is_authoritative':False})`
At the moment, the authoritative departments are pre-selected by default. If the user tries removing them, it does nothing. Do you still want the alternative of a read-only field displaying authoritative departments? I'd have to figure out how to do that.

Reply:
![image](https://github.com/ucb-rit/coldfront/assets/43729496/b0cf0c95-14cd-4234-b38d-9380bd69666f)

non-authoritative departments are preselected and authoritative ones are excluded now.